test/GatewayUpgradeFlow.js

```diff
@@ -0,0 +1,180 @@
+'use strict';
+
+const chai = require('chai');
+const { sortBy } = require('lodash');
+const { ethers, network } = require('hardhat');
+const {
+    utils: { defaultAbiCoder, Interface, solidityKeccak256 },
+} = ethers;
+const { expect } = chai;
+const { isHardhat, getAddresses, getWeightedAuthDeployParam, getWeightedProxyDeployParams } = require('./utils');
+const { getBytecodeHash } = require('@axelar-network/axelar-contract-deployments');
+
+const getWeights = ({ length }, weight = 1) => Array(length).fill(weight);
+
+describe('InterchainGovernance', () => {
+    const threshold = isHardhat ? 4 : 2;
+
+    let ownerWallet;
+    let governanceAddress;
+
+    let interchainGovernanceFactory;
+    let interchainGovernance;
+    let gatewayFactory;
+    let authFactory;
+    let gatewayProxyFactory;
+    let tokenDeployerFactory;
+
+    let auth;
+    let tokenDeployer;
+    let gateway;
+
+    let wallets;
+    let owner;
+    let operators;
+    let mintLimiter;
+
+    const governanceChain = 'Governance Chain';
+
+    before(async () => {
+        wallets = await ethers.getSigners();
+        ownerWallet = wallets[0];
+        mintLimiter = ownerWallet;
+        governanceAddress = wallets[1].address;
+        operators = sortBy(wallets.slice(0, threshold), (wallet) => wallet.address.toLowerCase());
+
+        interchainGovernanceFactory = await ethers.getContractFactory('TestInterchainGovernance', ownerWallet);
+
+        gatewayFactory = await ethers.getContractFactory('AxelarGateway', owner);
+        authFactory = await ethers.getContractFactory('AxelarAuthWeighted', owner);
+        gatewayProxyFactory = await ethers.getContractFactory('AxelarGatewayProxy', owner);
+        tokenDeployerFactory = await ethers.getContractFactory('TokenDeployer', owner);
+        tokenDeployer = await tokenDeployerFactory.deploy();
+        await tokenDeployer.deployTransaction.wait(network.config.confirmations);
+    });
+
+    beforeEach(async () => {
```

Comment:
beforeEach will cause a lot of txs when running on a live network.
Put this in a helper function and group the set of tests in more describe blocks. For e.g. all the queries below can be grouped together, and gateway can be deployed just once to test that (using a before block). See how AxelarGateway test does it

Reply:
Moved gateway deployment into helper function and reorganized tests